plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts

```diff
@@ -0,0 +1,268 @@
+  TaskSpec,
+  TaskStore,
+  TaskStoreEmitOptions,
+  TaskStoreGetEventsOptions,
+} from './types';
+
+const migrationsDir = resolvePackagePath(
+  '@backstage/plugin-scaffolder-backend',
+  'migrations',
+);
+
+export type RawDbTaskRow = {
+  id: string;
+  spec: string;
+  status: Status;
+  last_heartbeat_at?: string;
+  created_at: string;
+};
+
+export type RawDbTaskEventRow = {
+  id: number;
+  task_id: string;
+  body: string;
+  event_type: TaskEventType;
+  created_at: string;
+};
+
+export class DatabaseTaskStore implements TaskStore {
+  static async create(knex: Knex): Promise<DatabaseTaskStore> {
+    await knex.migrate.latest({
+      directory: migrationsDir,
+    });
+    return new DatabaseTaskStore(knex);
+  }
+
+  constructor(private readonly db: Knex) {}
+
+  async get(taskId: string): Promise<DbTaskRow> {
+    const [result] = await this.db<RawDbTaskRow>('tasks')
+      .where({ id: taskId })
+      .select();
+    if (!result) {
+      throw new NotFoundError(`No task with id '${taskId}' found`);
+    }
+    try {
+      const spec = JSON.parse(result.spec);
+      return {
+        id: result.id,
+        spec,
+        status: result.status,
+        lastHeartbeatAt: result.last_heartbeat_at,
+        createdAt: result.created_at,
+      };
+    } catch (error) {
+      throw new Error(`Failed to parse spec of task '${taskId}', ${error}`);
+    }
+  }
+
+  async createTask(spec: TaskSpec): Promise<{ taskId: string }> {
+    const taskId = uuid();
+    await this.db<RawDbTaskRow>('tasks').insert({
+      id: taskId,
+      spec: JSON.stringify(spec),
+      status: 'open',
+    });
+    return { taskId };
+  }
+
+  async claimTask(): Promise<DbTaskRow | undefined> {
+    return this.db.transaction(async tx => {
+      const [task] = await tx<RawDbTaskRow>('tasks')
+        .where({
+          status: 'open',
+        })
+        .limit(1)
+        .select();
+
+      if (!task) {
+        return undefined;
+      }
+
+      const updateCount = await tx<RawDbTaskRow>('tasks')
+        .where({ id: task.id, status: 'open' })
+        .update({
+          status: 'processing',
+          last_heartbeat_at: this.db.fn.now(),
+        });
+
+      if (updateCount < 1) {
+        return undefined;
+      }
+
+      try {
+        const spec = JSON.parse(task.spec);
+        return {
+          id: task.id,
+          spec,
+          status: 'processing',
+          lastHeartbeatAt: task.last_heartbeat_at,
+          createdAt: task.created_at,
+        };
+      } catch (error) {
+        throw new Error(`Failed to parse spec of task '${task.id}', ${error}`);
+      }
+    });
+  }
+
+  async heartbeatTask(taskId: string): Promise<void> {
+    const updateCount = await this.db<RawDbTaskRow>('tasks')
+      .where({ id: taskId, status: 'processing' })
+      .update({
+        last_heartbeat_at: this.db.fn.now(),
+      });
+    if (updateCount === 0) {
+      throw new Error(`No running task with taskId ${taskId} found`);
```

Comment:
should this be a ConflictError instead so the caller can get a clean HTTP error code for it?

Reply:
We all like clean HTTP errors, updated 👍